src/ext/legit/legit-commit.lisp

```diff
@@ -0,0 +1,125 @@
+(in-package :lem/legit)
+
+#|
+Done:
+
+- "c" opens a commit message window on the right side, we can type a long message,
+- lines starting by a # are ignored.
+
+TODOs:
+
+- add C-c C-s to sign the message
+- and other shortcuts and features
+
+Future:
+
+- save previous messages, add C-p C-n commands to show them
+- find related Github issue when writing "fixes #"
+
+|#
+
+(define-major-mode legit-commit-mode lem-markdown-mode:markdown-mode
+    (:name "legit-commit-mode"
+     :syntax-table lem-markdown-mode::*markdown-syntax-table*
+     :keymap *legit-commit-mode-keymap*)
+  ;; no syntax highlihgt in fact.
+  (setf (variable-value 'enable-syntax-highlight) t))
+
+;; Validate, abort.
+(define-key *legit-commit-mode-keymap* "C-c C-c" 'commit-continue)
+(define-key *legit-commit-mode-keymap* "C-Return" 'commit-continue)
+(define-key *legit-commit-mode-keymap* "M-q" 'commit-abort)
+(define-key *legit-commit-mode-keymap* "C-c C-k" 'commit-abort)
+
+;; Navigation.
+;; find and display the previous commit messages.
+;; (define-key *legit-commit-mode-keymap* "C-n" 'next-commit)
+;; (define-key *legit-commit-mode-keymap* "C-p" 'previous-commit)
+
+;; Help.
+(define-key *legit-commit-mode-keymap* "C-x ?" 'commit-help)
+
+(defun commit ()
+  (let ((buffer (make-buffer "*legit-commit*")))
+    (setf (buffer-directory buffer) (uiop:getcwd))
+    (setf (buffer-read-only-p buffer) nil)
+    (erase-buffer buffer)
+    (move-to-line (buffer-point buffer) 1)
+    (insert-string (buffer-point buffer) "write commit message:")
+    (change-buffer-mode buffer 'legit-commit-mode)
+    ;; (setf (buffer-read-only-p buffer) t)
+    (move-to-line (buffer-point buffer) 1)))
+
+(defun clean-commit-message (text)
+  "Remove lines starting with a #."
+  ;; We should collect meaningful data too, like a signature.
+  (loop for line in (str:lines text)
+        unless (str:starts-with-p "#" line)
+          collect line into result
+        finally (return (str:unlines result))))
+
+;; void message:
+#+(or)
+(assert (str:blankp (clean-commit-message
+"# Please enter the commit message for your changes. Lines starting
+# with '#' will be ignored, and an empty message aborts the commit.
+")))
+
+;; a message on the first line:
+#+(or)
+(assert (equal "test message" (clean-commit-message
+"test message
+# Please enter the commit message for your changes. Lines starting
+# with '#' will be ignored, and an empty message aborts the commit.
+")))
+
+;; a few lines:
+#+(or)
+(assert (equal "one
+
+two
+" (clean-commit-message
+"one
+
+two
+
+# Please enter the commit message for your changes. Lines starting
+# with '#' will be ignored, and an empty message aborts the commit.
+")))
+
+(defun commit-continue ()
```

Comment:
This is bound to C-c C-c, but it is not defined using define-command, so it cannot be called.